ray-operator/test/e2e/rayjob_recovery_test.go

```diff
@@ -0,0 +1,109 @@
+package e2e
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	rayv1 "github.com/ray-project/kuberay/ray-operator/apis/ray/v1"
+	rayv1ac "github.com/ray-project/kuberay/ray-operator/pkg/client/applyconfiguration/ray/v1"
+	. "github.com/ray-project/kuberay/ray-operator/test/support"
+)
+
+func TestRayJobRecovery(t *testing.T) {
+	test := With(t)
+	g := NewWithT(t)
+
+	// Create a namespace
+	namespace := test.NewTestNamespace()
+
+	// Job scripts
+	jobsAC := newConfigMap(namespace.Name, files(test, "long_running_counter.py"))
+	jobs, err := test.Client().Core().CoreV1().ConfigMaps(namespace.Name).Apply(test.Ctx(), jobsAC, TestApplyOptions)
+	g.Expect(err).NotTo(HaveOccurred())
+	test.T().Logf("Created ConfigMap %s/%s successfully", jobs.Namespace, jobs.Name)
+
+	test.T().Run("RayJob should recover after pod deletion", func(_ *testing.T) {
+		rayJobAC := rayv1ac.RayJob("counter", namespace.Name).
+			WithSpec(rayv1ac.RayJobSpec().
+				WithRayClusterSpec(newRayClusterSpec(mountConfigMap[rayv1ac.RayClusterSpecApplyConfiguration](jobs, "/home/ray/jobs"))).
+				WithEntrypoint("python /home/ray/jobs/long_running_counter.py").
+				WithRuntimeEnvYAML(`
+env_vars:
+  counter_name: test_counter
+`).
+				WithShutdownAfterJobFinishes(true).
+				WithSubmitterPodTemplate(jobSubmitterPodTemplateApplyConfiguration()))
+
+		rayJob, err := test.Client().Ray().RayV1().RayJobs(namespace.Name).Apply(test.Ctx(), rayJobAC, TestApplyOptions)
+		g.Expect(err).NotTo(HaveOccurred())
+		test.T().Logf("Created RayJob %s/%s successfully", rayJob.Namespace, rayJob.Name)
+
+		test.T().Logf("Waiting for RayJob %s/%s to start running", rayJob.Namespace, rayJob.Name)
+		g.Eventually(RayJob(test, rayJob.Namespace, rayJob.Name), TestTimeoutMedium).
+			Should(WithTransform(RayJobStatus, Equal(rayv1.JobStatusRunning)))
+		test.T().Logf("Find RayJob %s/%s running", rayJob.Namespace, rayJob.Name)
+		// wait for the job to run a bit
+		test.T().Logf("Sleep RayJob %s/%s 15 seconds", rayJob.Namespace, rayJob.Name)
+		time.Sleep(15 * time.Second)
+
+		// get the running jobpods
+		jobpods, err := test.Client().Core().CoreV1().Pods(namespace.Name).List(test.Ctx(), metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("job-name=%s", rayJob.Name),
+		})
+		g.Expect(err).NotTo(HaveOccurred())
+
+		// remove the running jobpods
+		propagationPolicy := metav1.DeletePropagationBackground
+		for _, pod := range jobpods.Items {
+			test.T().Logf("Delete Pod %s from namespace  %s", pod.Name, rayJob.Namespace)
+			err = test.Client().Core().CoreV1().Pods(namespace.Name).Delete(test.Ctx(), pod.Name, metav1.DeleteOptions{
+				PropagationPolicy: &propagationPolicy,
+			})
+			g.Expect(err).NotTo(HaveOccurred())
+		}
+
+		// Get old pod names for comparison
+		oldPodNames := make([]string, len(jobpods.Items))
+		for i, pod := range jobpods.Items {
+			oldPodNames[i] = pod.Name
+		}
```

Comment:
It seems that `oldPodNames` is not used below.

Reply:
Oh, I forgot I chose to use `jobpods.Items` directly. Thank you!